### PR TITLE
Deprecate useBlockEditContext in favour of useBlockClientId

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

Ideally the only piece of context is the block client ID and everything else is derived from that. Unfortunately `useBlockEditContext` is exported at the package level (not sure why because it's only used within the package), but it can be rewritten with `useSelect`.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/block-editor/src/components/block-edit/context.js [hunk 1]
```diff
@@ -1,29 +1,46 @@
...
  * WordPress dependencies
  */
+import { useSelect } from '@wordpress/data';
 import { createContext, useContext } from '@wordpress/element';
 
-const Context = createContext( {
-	name: '',
-	isSelected: false,
-	focusedElement: null,
-	setFocusedElement: noop,
```
Comment (ellatrix) [inline packages/block-editor/src/components/block-edit/context.js:15]: `focusedElement` and `setFocusedElement` is context that has been removed a long time ago.

**Comment by ellatrix** [comment]
Not sure where to failures are coming from so splitting into smaller PRs to pinpoint.

**Comment by fluiddot** [comment]
👋 I think I've found the reason of the failure of the `Unit Tests / Mobile` check, for some reason the prop `isSelected` of the [RichText component (native version)](https://github.com/WordPress/gutenberg/blob/master/packages/rich-text/src/component/index.native.js#L731) with these changes is now `true` when the component is mounted, however with changes from `master` the prop's value is `false`.

I'm not sure why this prop is now receiving the opposite value but it would be nice to investigate it further just in case it produces any undesired side effects.

Regarding the failure, this uncovered a problem we have in the unit tests because the `focus` function of the `AztecView` component is now being [called](https://github.com/WordPress/gutenberg/blob/master/packages/rich-text/src/component/index.native.js#L734) due to to having the prop `isSelected` as `true`. That function should be mocked because it calls native functions not available when running unit tests, but unfortunately it's not and therefore when calling native functions it fails.

I'm going to open a PR adding the proper mock for the `AztecView` component but I'll keep investigating potential side effects.

**Comment by fluiddot** [comment]
Upon further investigation I realised that the root cause of all the issues is related to how we calculate if a block is selected, specifically for the `RichText` component. 

Previous to these changes, [the value was passed through the block edit context](https://github.com/WordPress/gutenberg/pull/29309/files#diff-bec072c72a2f961596701aef67f67575b0a3b918135a6587229a0cd79bb3db44L158) which is not being used when running unit tests so that value was always `undefined`. 

[Now as the value is fetched directly via the selector](https://github.com/WordPress/gutenberg/pull/29309/files#diff-bec072c72a2f961596701aef67f67575b0a3b918135a6587229a0cd79bb3db44R202), it's returning a different value. The cause of this is that the blocks rendered within unit tests don't have a `clientId` and the `isBlockSelected` selector is not properly calculating the value. In fact it's returning `true` because it does a comparison of the state's value and the one passed (`undefined === undefined`) which is not correct ([related code](https://github.com/WordPress/gutenberg/blob/try/block-context-cleanup/packages/block-editor/src/store/selectors.js#L967)). 

It would be nice if we add this case to the selector and check if the `clientId` passed is `undefined` and return `false` in that case.

Apart from that I think the changes are safe from the mobile version perspective 🎊 .

**Comment by fluiddot** [comment]
> Upon further investigation I realised that the root cause of all the issues is related to how we calculate if a block is selected, specifically for the `RichText` component.
> 
> Previous to these changes, [the value was passed through the block edit context](https://github.com/WordPress/gutenberg/pull/29309/files#diff-bec072c72a2f961596701aef67f67575b0a3b918135a6587229a0cd79bb3db44L158) which is not being used when running unit tests so that value was always `undefined`.
> 
> [Now as the value is fetched directly via the selector](https://github.com/WordPress/gutenberg/pull/29309/files#diff-bec072c72a2f961596701aef67f67575b0a3b918135a6587229a0cd79bb3db44R202), it's returning a different value. The cause of this is that the blocks rendered within unit tests don't have a `clientId` and the `isBlockSelected` selector is not properly calculating the value. In fact it's returning `true` because it does a comparison of the state's value and the one passed (`undefined === undefined`) which is not correct ([related code](https://github.com/WordPress/gutenberg/blob/try/block-context-cleanup/packages/block-editor/src/store/selectors.js#L967)).
> 
> It would be nice if we add this case to the selector and check if the `clientId` passed is `undefined` and return `false` in that case.
> 
> Apart from that I think the changes are safe from the mobile version perspective 🎊 .

I created this [PR](https://github.com/WordPress/gutenberg/pull/29377) that fixes the mentioned issue.

**Comment by fluiddot** [comment]
👋 I created this [PR](https://github.com/WordPress/gutenberg/pull/29569) that should fix the errors when running the native unit tests of this PR.

My approach, as described in the PR, is to basically render the `BlockEdit` component in the unit tests instead of the edit component. This way we add the context that is required by some blocks.

**Comment by ellatrix** [comment]
@fluiddot I got rid of the problem by returning early in case there is no client ID: https://github.com/WordPress/gutenberg/commit/f0a0056430f16e1f81e06d45534bd8db3c320074#diff-d5252e488c7dc0d44aba3d01d607ba2d77100d56d7513d82fc3839142ecf6347R27

**Comment by fluiddot** [comment]
> @fluiddot I got rid of the problem by returning early in case there is no client ID: [f0a0056#diff-d5252e488c7dc0d44aba3d01d607ba2d77100d56d7513d82fc3839142ecf6347R27](https://github.com/WordPress/gutenberg/commit/f0a0056430f16e1f81e06d45534bd8db3c320074#diff-d5252e488c7dc0d44aba3d01d607ba2d77100d56d7513d82fc3839142ecf6347R27)

I'm afraid that won't fix all the errors because the problem is related to not having a defined `clientID`. For example, [one of the errors from the unit tests](https://github.com/WordPress/gutenberg/pull/29309/checks?check_run_id=2035496869) is caused by the `RichText` editor component, that is receiving `true` when calling the `isBlockSelected` selector ([related code](https://github.com/WordPress/gutenberg/pull/29309/files#diff-bec072c72a2f961596701aef67f67575b0a3b918135a6587229a0cd79bb3db44R215)).
 
The [PR](https://github.com/WordPress/gutenberg/pull/29569) I made for fixing this is ready to merge so once the PR check passes, I'll merge it and you could include the changes in this branch.

**Pull request closed without merging**